### ISSUE TITLE
xvector-vits conversion error

### ISSUE THREAD

**Issue by unparalleled-ysj** [Masao-Someki/espnet_onnx#37]
Hi @Masao-Someki , have you tried converting xvector-vits to onnx ? when I try to do this, I run into this problem:

/root/anaconda3/envs/espnet/lib/python3.7/site-packages/torch/onnx/symbolic_helper.py:719: UserWarning: allowzero=0 by default. In order to honor zero value in shape use allowzero=1
  warnings.warn("allowzero=0 by default. In order to honor zero value in shape use allowzero=1")
WARNING: The shape inference of prim::Constant type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Floating point exception(core dump)

**Comment by Masao-Someki** [comment]
@unparalleled-ysj 
Could you tell me the package versions? The result of the `pip list` is enough.
`onnx==1.12.0` might cause this error... If you use v1.12.0, would you check it with v1.11.0?

**Comment by unparalleled-ysj** [comment]
> @unparalleled-ysj 
> 
> Could you tell me the package versions? The result of the `pip list` is enough.
> 
> `onnx==1.12.0` might cause this error... If you use v1.12.0, would you check it with v1.11.0?

I use onnx==1.11.0

**Comment by Masao-Someki** [comment]
Hmm...
Sorry, I have never tried xvector-vits before. I think there might be a bug somewhere in `spemb` exportation. Currently, I don't have much time to further investigate on this, so is it okay to debug this weekend?

**Comment by unparalleled-ysj** [comment]
of cause, thank you for your support！

**Comment by unparalleled-ysj** [comment]
@Masao-Someki Hi, I tried to debug by myself and found that the problem occurs in [https://github.com/Masao-Someki/espnet_onnx/blob/a2921fc2bd42ce468994dac49fa610729b056623/espnet_onnx/export/tts/models/tts_models/vits.py#L105](F.normalize()), when I removed the normalize operation, the xvector-vits model was successfully exported. So are we going to implement a normalize operation ourselves?

**Comment by Masao-Someki** [comment]
@unparalleled-ysj 
Thank you!
I see that the `eps` for the `F.noramlize` in `1e-12`, which I think is too small. In espnet_onnx, I set `-10000.0` as the minimum value when computing some operations like softmax.
I'm not testing this, but maybe this could be the cause of the exception?

**Comment by unparalleled-ysj** [comment]
@Masao-Someki Actually, F.normalize is not imported in https://github.com/Masao-Someki/espnet_onnx/blob/a2921fc2bd42ce468994dac49fa610729b056623/espnet_onnx/export/tts/models/tts_models/vits.py#L105 , so I changed it to nn.functional.normalize,  eps=1e-12 and eps=-10000.0 I have tried both, but there will cause the exception.

**Comment by unparalleled-ysj** [comment]
After ablation experiments, it was finally determined that the clamp_min(eps) operation in F.noramilize caused floating point exception

def normalize(input: Tensor, p: float = 2.0, dim: int = 1, eps: float = 1e-12, out: Optional[Tensor] = None) -> Tensor:
    r"""Performs :math:`L_p` normalization of inputs over specified dimension.

    For a tensor :attr:`input` of sizes :math:`(n_0, ..., n_{dim}, ..., n_k)`, each
    :math:`n_{dim}` -element vector :math:`v` along dimension :attr:`dim` is transformed as

    .. math::
        v = \frac{v}{\max(\lVert v \rVert_p, \epsilon)}.

    With the default arguments it uses the Euclidean norm over vectors along dimension :math:`1` for normalization.

    Args:
        input: input tensor of any shape
        p (float): the exponent value in the norm formulation. Default: 2
        dim (int): the dimension to reduce. Default: 1
        eps (float): small value to avoid division by zero. Default: 1e-12
        out (Tensor, optional): the output tensor. If :attr:`out` is used, this
                                operation won't be differentiable.
    """
    if has_torch_function_variadic(input, out):
        return handle_torch_function(normalize, (input, out), input, p=p, dim=dim, eps=eps, out=out)
    if out is None:
        denom = input.norm(p, dim, keepdim=True).clamp_min(eps).expand_as(input)
        return input / denom
    else:
        denom = input.norm(p, dim, keepdim=True).clamp_min_(eps).expand_as(input)
        return torch.div(input, denom, out=out)

**Comment by Masao-Someki** [comment]
@unparalleled-ysj Thank you for the information!
And I realized that I told you some super weird thing about eps... :scream: 

When the exported model does not use its input while inference about the input names, that input would be removed.
So I think that even the `spemb` is defined in `get_input_names` and `get_dummy_input`, it is not used in the model inference.
I used `spk_embed_dim` to define whether the model uses `spemb` or not, so maybe changing this line will take effect. Perhaps we should change the condition into something other, like `if spemb is not None`
https://github.com/Masao-Someki/espnet_onnx/blob/a2921fc2bd42ce468994dac49fa610729b056623/espnet_onnx/export/tts/models/tts_models/vits.py#L103

**Comment by unparalleled-ysj** [comment]
> @unparalleled-ysj Thank you for the information! And I realized that I told you some super weird thing about eps... 😱
> 
> When the exported model does not use its input while inference about the input names, that input would be removed. So I think that even the `spemb` is defined in `get_input_names` and `get_dummy_input`, it is not used in the model inference. I used `spk_embed_dim` to define whether the model uses `spemb` or not, so maybe changing this line will take effect. Perhaps we should change the condition into something other, like `if spemb is not None`
> 
> https://github.com/Masao-Someki/espnet_onnx/blob/a2921fc2bd42ce468994dac49fa610729b056623/espnet_onnx/export/tts/models/tts_models/vits.py#L103

According to your suggestion, I tried it and it doesn't work,  and I think even if spk_embed_dim  define whether the model uses spemb or no, in this line , spembs  has been used during model inference.
https://github.com/Masao-Someki/espnet_onnx/blob/a2921fc2bd42ce468994dac49fa610729b056623/espnet_onnx/export/tts/models/tts_models/vits.py#L106

Strangely, even though onnx_model.get_inputs() is ['text', 'feats'], 'feats' plays the role of spembs, so I think it's an input naming problem with torch.onnx.export()

**Comment by unparalleled-ysj** [comment]
Bug solved, thank you @Masao-Someki  for your support
